contracts/rate_oracles/CompoundRateOracle.sol

```diff
@@ -0,0 +1,255 @@
+    /// @dev Reverts if we have no data point for either timestamp
+    /// @param _from The timestamp of the start of the period, in seconds
+    /// @param _to The timestamp of the end of the period, in seconds
+    /// @return The "floating rate" expressed in Wad, e.g. 4% is encoded as 0.04*10**18 = 4*10**16
+    function getRateFromTo(
+        uint256 _from,
+        uint256 _to //  move docs to IRateOracle. Add additional parameter to use cache and implement cache.
+    ) public view override(BaseRateOracle, IRateOracle) returns (uint256) {
+        require(_from <= _to, "from > to");
+
+        if (_from == _to) {
+            return 0;
+        }
+
+        // note that we have to convert comp index into "floating rate" for
+        // swap calculations, e.g. an index multiple of 1.04*10**27 corresponds to
+        // 0.04*10**27 = 4*10*25
+        uint32 currentTime = Time.blockTimestampTruncated();
+        uint32 from = Time.timestampAsUint32(_from);
+        uint32 to = Time.timestampAsUint32(_to);
+
+        uint256 rateFromRay = observeSingle(
+            currentTime,
+            from,
+            oracleVars.rateIndex,
+            oracleVars.rateCardinality
+        );
+        uint256 rateToRay = observeSingle(
+            currentTime,
+            to,
+            oracleVars.rateIndex,
+            oracleVars.rateCardinality
+        );
+
+        if (rateToRay > rateFromRay) {
+            uint256 result = WadRayMath.rayToWad(
+                WadRayMath.rayDiv(rateToRay, rateFromRay) - WadRayMath.RAY
+            );
+            return result;
+        } else {
+            return 0;
+        }
+    }
+
+    /// @notice Calculates the interpolated (counterfactual) rate value
+    /// @param beforeOrAtRateValueRay  Rate Value (in ray) before the timestamp for which we want to calculate the counterfactual rate value
+    /// @param apyFromBeforeOrAtToAtOrAfterWad Apy in the period between the timestamp of the beforeOrAt Rate and the atOrAfter Rate
+    /// @param timeDeltaBeforeOrAtToQueriedTimeWad Time Delta (in wei seconds) between the timestamp of the beforeOrAt Rate and the atOrAfter Rate
+    /// @return rateValueRay Counterfactual (interpolated) rate value in ray
+    /// @dev Given [beforeOrAt, atOrAfter] where the timestamp for which the counterfactual is calculated is within that range (but does not touch any of the bounds)
+    /// @dev We can calculate the apy for [beforeOrAt, atOrAfter] --> refer to this value as apyFromBeforeOrAtToAtOrAfter
+    /// @dev Then we want a counterfactual rate value which results in apy_before_after if the apy is calculated between [beforeOrAt, timestampForCounterfactual]
+    /// @dev Hence (1+rateValueWei/beforeOrAtRateValueWei)^(1/timeInYears) = apyFromBeforeOrAtToAtOrAfter
+    /// @dev Hence rateValueWei = beforeOrAtRateValueWei * (1+apyFromBeforeOrAtToAtOrAfter)^timeInYears - 1)
+    function interpolateRateValue(
```

Comment:
is this function a duplicate of the one in aave rate oracle, if yes, can push into a library?

Reply:
Done